sdk/resourcemanager/Azure.ResourceManager.Core/src/Resources/ResourceIdentifier.cs

```diff
@@ -3,295 +3,439 @@
+
+        internal const string ResourceGroupsLowerKey = "resourcegroups"; 
+
+        internal const string BuiltInResourceNamespace = "Microsoft.Resources";
+
+        internal static ResourceType SubscriptionType => new ResourceType(BuiltInResourceNamespace, SubscriptionsKey);
+        internal static ResourceType LocationsType =>
+            new ResourceType(BuiltInResourceNamespace, $"{SubscriptionsKey}/{LocationsKey}");
+        internal static ResourceType ResourceGroupsType =>
+            new ResourceType(BuiltInResourceNamespace, $"{SubscriptionsKey}/{ResourceGroupsKey}");
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="ResourceIdentifier"/> class.
+        /// The root of the resource hierarchy
+        /// </summary>
+        public static ResourceIdentifier RootResourceIdentifier => new RootResourceIdentifier();
+
+        /// <summary>
+        /// For internal use only.
         /// </summary>
-        /// <param name="id"> The identifier of the resource that is the target of operations. </param>
-        public ResourceIdentifier(string id)
+        internal ResourceIdentifier()
         {
-            Id = id;
-            Parse(id);
+            _stringValue = null;
         }
 
         /// <summary>
-        /// Gets the Resource ID.
+        /// For internal use only.
         /// </summary>
-        public string Id { get; private set; }
+        /// <param name="resourceType"> The type of the resource.</param>
+        /// <param name="name"> The name of the resource.</param>
+        internal ResourceIdentifier(ResourceType resourceType, string name)
+        {
+            ResourceType = resourceType;
+            Name = name;
+            _stringValue = null;
+        }
 
         /// <summary>
-        /// Gets the Resource Name.
+        /// Initializes a new instance of the <see cref="ResourceIdentifier"/> class.
         /// </summary>
-        public string Name { get; private set; }
+        /// <param name="resourceId"> The string representation of a resource Id. </param>
+        /// <returns> The resource identifier. </returns>
+        public static ResourceIdentifier Create(string resourceId)
+        {
+            if (resourceId is null)
+                throw new ArgumentNullException(nameof(resourceId));
+            if (!resourceId.StartsWith("/", StringComparison.InvariantCultureIgnoreCase))
+                throw new ArgumentOutOfRangeException(nameof(resourceId), "Invalid resource id.");
+            var parts = resourceId.Split(new[] { '/' }, StringSplitOptions.RemoveEmptyEntries).ToList();
+            if (parts.Count < 2)
+                throw new ArgumentOutOfRangeException(nameof(resourceId), "Invalid resource id.");
+            switch (parts[0].ToLowerInvariant())
+            {
+                case SubscriptionsKey:
+                    return CreateBaseSubscriptionIdentifier(parts[1], parts.Trim(2));
```

Comment:
should we be keeping the raw string around here?  It seems we might be introducing some responsibility that didn't exist before.

The service returns string X
we convert string X into Parts[]
later we use a string builder to combine those back into Y

How confident are we that Y will always equal X when X is valid of course and passes validation

Reply:
I don't see any compelling value from retaining the input string, but perhaps there is a scenario you have in mind that I have not considered